app/ts/store/account.ts

```diff
@@ -1,77 +1,89 @@
 import { ethers } from 'ethers'
-import { isEthereumJsonRpcError, EthereumJsonRpcError } from '../library/exceptions.js'
+import { AsyncProperty, useAsyncState } from '../library/preact-utilities.js'
 import { assertsExternalProvider, isEthereumObservable } from '../library/utilities.js'
 
-type AccountBusy = {
-	status: 'busy'
+export type ConnectMutation = {
+	transport: Signal<AsyncProperty<string>>
+	dispatch: () => void
+	reset: () => void
 }
 
-type AccountConnected = {
-	status: 'connected'
-	address: string
+export type ReconnectMutation = {
+	transport: Signal<AsyncProperty<unknown>>
+	dispatch: () => void
+	reset: () => void
 }
 
-type AccountDisconnected = {
-	status: 'disconnected'
-	connect: () => Promise<void>
-	ensureConnected: () => Promise<void>
-}
+type Account =
+	| {
+			isConnected: false
+			connectMutation: ConnectMutation
+			reconnectMutation: ReconnectMutation
+	  }
+	| {
+			isConnected: true
+			address: string
+	  }
 
-type AccountConnectRejected = {
-	status: 'rejected'
-	error: EthereumJsonRpcError | Error
-}
+export type AccountStore = ReturnType<typeof createAccountStore>
+export function createAccountStore() {
+	const { value: asyncValue, waitFor, reset } = useAsyncState<string>()
+	const reconnect = useAsyncState()
+
+	const reconnectMutation = {
+		transport: reconnect.value,
+		dispatch: () => {
+			reconnect.waitFor(async () => {
+				try {
+					assertsExternalProvider(window.ethereum)
+					const provider = new ethers.providers.Web3Provider(window.ethereum)
+					const signer = provider.getSigner()
+					const address = await signer.getAddress()
+					accountStore.value = { isConnected: true, address }
+				} catch {
+					accountStore.value = accountDefaults
+				}
+			})
+		},
+		reset: reconnect.reset,
+	}
 
-export type AccountStore = AccountBusy | AccountConnected | AccountDisconnected | AccountConnectRejected
+	const connectMutation = {
+		transport: asyncValue,
+		dispatch: () => {
+			waitFor(async () => {
+				assertsExternalProvider(window.ethereum)
+				const provider = new ethers.providers.Web3Provider(window.ethereum)
+				await provider.send('eth_requestAccounts', [])
+				const signer = provider.getSigner()
+				const address = await signer.getAddress()
+				accountStore.value = { isConnected: true, address }
+				return address
+			})
+		},
+		reset,
+	}
+
+	const accountDefaults = { isConnected: false as const, connectMutation, reconnectMutation }
+	const accountStore = useSignal<Account>(accountDefaults)
+
+	const handleAccountChange = (newAccount: string[]) => {
+		if (ethers.utils.isAddress(newAccount[0])) {
+			accountStore.value = { isConnected: true, address: newAccount[0] }
+			return
+		}
 
-const storeDefaults = { status: 'disconnected', connect, ensureConnected } as const
-export const accountStore = signal<AccountStore>(storeDefaults)
+		reset()
+		accountStore.value = accountDefaults
+	}
 
-async function connect() {
-	try {
+	effect(() => {
```

Comment:
Why is this inside an `effect` rather just inline?  Do you perhaps want `effect(..., [])` so it only runs once per re-render of anything that uses it?

Reply:
I'm not aware `effect` accepting a second parameter looking at it's type, tho you are correct in assuming i'd only need to fire this once upon mount, so I suppose preact's `useEffect` would suffice.

Reply:
Yeah, I was thinking of `useEffect`, which would limit the frequency this is called I believe.